services/basic_types.proto

```diff
@@ -1493,3 +1493,67 @@ message TokenAssociation {
     TokenID token_id = 1;  // The token involved in the association
     AccountID account_id = 2; // The account involved in the association
 }
+
+/**
+ * An approved allowance for hbar transfers for a spender.
+ */
+message CryptoApproval {
+    /**
+     * The account ID of the spender of the hbar allowance.
+     */
+    AccountID spender = 1;
+
+    /**
+     * The amount of the spender's allowance in tinybars.
+     */
+    int64 amount = 2;
+}
+
+/**
+ * An approved allowance for fungible token transfers for a spender.
+ */
+message TokenApproval {
+    /**
+     * The token that the allowance pertains to.
+     */
+    TokenID tokenId = 1;
+
+    /**
+     * The account ID of the token allowance spender.
+     */
+    AccountID spender = 2;
+
+    /**
+     * The amount of the spender's token allowance.
+     */
+    int64 amount = 3;
+}
+
+/**
+ * An approved allowance for non-fungible token transfers for a spender.
+ */
+message NftApproval {
+    /**
+     * The token that the allowance pertains to.
+     */
+    TokenID tokenId = 1;
+
+    /**
+     * The account ID of the token allowance spender.
+     */
+    AccountID spender = 2;
+
+    oneof allowance {
+        /**
+         * The list of serial numbers that the spender is permitted to transfer.
+         */
+        repeated int64 serialNumbers = 3 [packed = true];
```

Comment:
For adjust, I assume all positive numbers are added to the list while discarding already existing serials?

And conversely for negative numbers they are removed from the list, but what happens if they don't exist?

What happens if `serialNumbers` contains duplicates?

Reply:
Yes, positive will add + negative will remove. It will be an error if the user tries to remove an NFT that the owner doesn't own. Before this change, duplicate token allowances in the list would be treated as an error since it would be confusing to know which one to pick if the amounts were different (do we take the last in the list?). Now that we have NFTs broken out, it's actually not harmful nor confusing if dups are in the list. It's really a waste of bytes and CPU to parse the protobuf. If it's easier for mirror node to not have an NFT list with dups, we'll return an error if we catch any.

Reply:
Would probably be best to error if duplicates.